Get hardware keyboard state off the main thread

pvollan opened the pull request:
#### 1ce83d6be586ef058792e65ad2895fafc05e3e2a
<pre>
Get hardware keyboard state off the main thread
<a href="https://bugs.webkit.org/show_bug.cgi?id=278885">https://bugs.webkit.org/show_bug.cgi?id=278885</a>
<a href="https://rdar.apple.com/133245137">rdar://133245137</a>

Reviewed by Chris Dumez.

In order to avoid blocking the main thread, get the hardware keyboard state off the main thread.
This is being done in the Web process pool, which holds the cached value for the keyboard state.

* Source/WebKit/UIProcess/API/Cocoa/WKWebView.mm:
(-[WKWebView _initializeWithConfiguration:]):
* Source/WebKit/UIProcess/Cocoa/WebProcessPoolCocoa.mm:
(WebKit::WebProcessPool::platformInitialize):
(WebKit::WebProcessPool::cachedHardwareKeyboardState):
(WebKit::WebProcessPool::setCachedHardwareKeyboardState):
* Source/WebKit/UIProcess/WebPageProxy.cpp:
(WebKit::WebPageProxy::creationParameters):
* Source/WebKit/UIProcess/WebPageProxyInternals.h:
* Source/WebKit/UIProcess/WebProcessPool.h:
* Source/WebKit/UIProcess/ios/WebPageProxyIOS.mm:
(WebKit::WebPageProxy::platformInitialize):
(WebKit::WebPageProxy::hardwareKeyboardAvailabilityChanged):

Canonical link: <a href="https://commits.webkit.org/283035@main">https://commits.webkit.org/283035@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/6d73a6541dcd960414603bafd6fe3ab1d49684a5

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/64986 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/44353 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/17600 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/69010 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/59/builds/15592 "Built successfully") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/52136 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/61/builds/15874 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/52205 "Passed tests") | [✅ 🧪 wincairo-tests](https://ews-build.webkit.org/#/builders/60/builds/10765 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/68052 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/41006 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/56241 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/32827 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/37677 "Passed tests") | | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/14468 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/59589 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/13950 "Passed tests") | [❌ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/70715 "Hash 6d73a654 for PR 32909 does not build (failure)") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/87/builds/8938 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/62/builds/13430 "Passed tests") | [❌ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/2/builds/70715 "Hash 6d73a654 for PR 32909 does not build (failure)") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/86/builds/8970 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/56301 "Passed tests") | [❌ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/2/builds/70715 "Hash 6d73a654 for PR 32909 does not build (failure)") | 
| | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/88/builds/7372 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/119/builds/1046 "Passed tests") | | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/9854 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/40165 "Built successfully") | | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/41242 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/42423 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/40986 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

pxlcoder commented on Source/WebKit/UIProcess/Cocoa/WebProcessPoolCocoa.mm:
```diff
@@ -1269,4 +1288,18 @@ static void webProcessPoolHighDynamicRangeDidChangeCallback(CFNotificationCenter
+void WebProcessPool::setCachedHardwareKeyboardState(HardwareKeyboardState hardwareKeyboardState)
+{
+    RELEASE_ASSERT(isMainRunLoop());
+    m_hardwareKeyboardState = hardwareKeyboardState;
```
What updates this cached value if hardware keyboard availability changes?

pvollan replied:
Ah, good catch, I will update this patch with an observer of this state.

Thanks for reviewing!

pxlcoder commented on Source/WebKit/UIProcess/API/Cocoa/WKWebView.mm:
```diff
@@ -329,17 +329,6 @@ static uint32_t convertSystemLayoutDirection(NSUserInterfaceLayoutDirection dire
-    RetainPtr webView { (__bridge WKWebView *)observer };
-    if (!webView)
-        return;
-    webView->_page->hardwareKeyboardAvailabilityChanged();
```
I think this patch introduces a subtle behavior change due to the removal of this call.

It seems that nothing calls `-[WKContentView _hardwareKeyboardAvailabilityChanged]` anymore, which is responsible for reloading input views.

pvollan replied:
Ah, good point, will address this in the upcoming patch.

Thanks for reviewing!

cdumez commented on Source/WebKit/UIProcess/ios/WebPageProxyIOS.mm:
```diff
@@ -1151,10 +1150,6 @@ static FloatSize fullscreenPreferencesScreenSize(CGFloat preferredWidth)
 void WebPageProxy::hardwareKeyboardAvailabilityChanged()
 {
     auto hardwareKeyboardState = currentHardwareKeyboardState();
-    if (hardwareKeyboardState == internals().hardwareKeyboardState)
```
If I remember correctly @pxlcoder had added this check to fix a critical bug where we ended with with an infinite notification loop between the WebProcess posting the notification, the UIProcess receiving it and sending IPC to the WebProcess causing it to notify again. Is this safe now to drop?

pvollan replied:
That's a good point! Since the notification service is blocked in the WebContent process now, this should not be an issue anymore.

Thanks for reviewing!

cdumez commented on Source/WebKit/UIProcess/WebPageProxy.cpp:
```diff
@@ -10558,7 +10558,7 @@ WebPageCreationParameters WebPageProxy::creationParameters(WebProcessProxy& proc
     parameters.mimeTypesWithCustomContentProviders = protectedPageClient()->mimeTypesWithCustomContentProviders();
     parameters.deviceOrientation = m_deviceOrientation;
-    parameters.hardwareKeyboardState = internals().hardwareKeyboardState;
+    parameters.hardwareKeyboardState = m_legacyMainFrameProcess->processPool().cachedHardwareKeyboardState();
```
Seems there is no guarantee that the cached hardware keyboard state has been initialized yet, right? Since it's happening on a background thread. I guess that's fine since we'll send IPC to the WebProcess once we actually cache the value the first time to let it know the value "changed"? Hopefully that's what's going on but I thought I would point it out so you make sure.

pvollan replied:
That is correct. This patch relies on sending the keyboard state over to the WebContent process, once it is available.

cdumez commented on Source/WebKit/UIProcess/Cocoa/WebProcessPoolCocoa.mm:
```diff
@@ -304,6 +307,19 @@ static void logProcessPoolState(const WebProcessPool& pool)
 void WebProcessPool::platformInitialize(NeedsGlobalStaticInitialization needsGlobalStaticInitialization)
 {
+#if PLATFORM(IOS_FAMILY)
+    dispatch_async(dispatch_get_global_queue(DISPATCH_QUEUE_PRIORITY_DEFAULT, 0), makeBlockPtr([weakThis = WeakPtr { *this }] {
```
Maybe we can move this to a separate function.

cdumez commented on Source/WebKit/UIProcess/Cocoa/WebProcessPoolCocoa.mm:
```diff
@@ -840,6 +877,9 @@ static void logProcessPoolState(const WebProcessPool& pool)
 
 #if PLATFORM(IOS_FAMILY)
+    auto notificationName = adoptNS([[NSString alloc] initWithCString:kGSEventHardwareKeyboardAvailabilityChangedNotification encoding:NSUTF8StringEncoding]);
+    addCFNotificationObserver(hardwareKeyboardAvailabilityChangedCallback, (__bridge CFStringRef)notificationName.get(), CFNotificationCenterGetDarwinNotifyCenter());
```
Curious if we could use `CFSTR(kGSEventHardwareKeyboardAvailabilityChangedNotification)` here

pvollan replied:
Ah, good point, will look into that!

pvollan replied:
I am getting a compile error with `CFSTR(kGSEventHardwareKeyboardAvailabilityChangedNotification`.

cdumez commented on Source/WebKit/UIProcess/Cocoa/WebProcessPoolCocoa.mm:
```diff
@@ -1269,4 +1309,18 @@ static void webProcessPoolHighDynamicRangeDidChangeCallback(CFNotificationCenter
 #endif
 
+#if PLATFORM(IOS_FAMILY)
+HardwareKeyboardState WebProcessPool::cachedHardwareKeyboardState()
```
Seems like this function should be `const`

cdumez commented on Source/WebKit/UIProcess/WebProcessPool.h:
```diff
@@ -679,6 +687,12 @@ class WebProcessPool final
     void clearAudibleActivity();
 
+#if PLATFORM(IOS_FAMILY)
+    static void hardwareKeyboardAvailabilityChangedCallback(CFNotificationCenterRef, void* observer, CFStringRef, const void*, CFDictionaryRef);
```
Does this need to be in the header?

pvollan replied:
It is not required, although I added this in the header to avoid making some private methods public.

cdumez replied:
Oh, it's fine then.

cdumez commented on Source/WebKit/UIProcess/Cocoa/WebProcessPoolCocoa.mm:
```diff
@@ -926,6 +971,8 @@ static void logProcessPoolState(const WebProcessPool& pool)
     [[NSNotificationCenter defaultCenter] removeObserver:m_accessibilityEnabledObserver.get()];
     [[NSNotificationCenter defaultCenter] removeObserver:m_applicationLaunchObserver.get()];
+    auto notificationName = adoptNS([[NSString alloc] initWithCString:kGSEventHardwareKeyboardAvailabilityChangedNotification encoding:NSUTF8StringEncoding]);
+    removeCFNotificationObserver((__bridge CFStringRef)notificationName.get());
```
How does this work when there is more than 1 process pool?

pvollan replied:
This call will only remove the observer for the calling process pool. Other process pools will still be observing this notification until they call this function. This is because each process pool is calling `CFNotificationCenterAddObserver` with a unique observer parameter.

cdumez replied:
Oh, there is logic in `removeCFNotificationObserver()`.. We weren't passing a pointer to `this` so I was confused.

pvollan committed:
Get hardware keyboard state off the main thread
https://bugs.webkit.org/show_bug.cgi?id=278885
rdar://133245137

Reviewed by Chris Dumez.

In order to avoid blocking the main thread, get the hardware keyboard state off the main thread.
This is being done in the Web process pool, which holds the cached value for the keyboard state.

* Source/WebKit/UIProcess/API/Cocoa/WKWebView.mm:
(-[WKWebView _initializeWithConfiguration:]):
* Source/WebKit/UIProcess/Cocoa/WebProcessPoolCocoa.mm:
(WebKit::WebProcessPool::platformInitialize):
(WebKit::WebProcessPool::cachedHardwareKeyboardState):
(WebKit::WebProcessPool::setCachedHardwareKeyboardState):
* Source/WebKit/UIProcess/WebPageProxy.cpp:
(WebKit::WebPageProxy::creationParameters):
* Source/WebKit/UIProcess/WebPageProxyInternals.h:
* Source/WebKit/UIProcess/WebProcessPool.h:
* Source/WebKit/UIProcess/ios/WebPageProxyIOS.mm:
(WebKit::WebPageProxy::platformInitialize):
(WebKit::WebPageProxy::hardwareKeyboardAvailabilityChanged):

Canonical link: https://commits.webkit.org/283035@main
